src/context/TodoContext.tsx

```diff
@@ -0,0 +1,139 @@
+  deleteTodo: (todoId: number) => void;
+  updateTodo: (todo: Omit<Todo, 'userId'>) => void;
+  handleSetTempTodo: (newTodo: Todo | null) => void;
+  handleSetErrorMessage: (newError: Error) => void;
+  handleSetUpdatingIds: (id: number | null) => void;
+  filterChange: (newStatus: Status) => void;
+};
+
+export const TodoContext = React.createContext<ContextProps>({
+  todos: [],
+  filteredTodos: [],
+  filter: Status.all,
+  tempTodo: null,
+  errorMessage: Error.none,
+  updatingIds: [],
+  addTodo: () => {},
+  deleteTodo: () => {},
+  updateTodo: () => {},
+  handleSetTempTodo: () => {},
+  handleSetErrorMessage: () => {},
+  handleSetUpdatingIds: () => {},
+  filterChange: () => {},
+});
+
+type Props = {
+  children: React.ReactNode;
+};
+
+export const GlobalProvider: React.FC<Props> = ({ children }) => {
+  const [todos, setTodos] = useState<Todo[]>([]);
+  const [errorMessage, setErrorMessage] = useState<Error | null>(null);
+  const [filter, setFilter] = useState(Status.all);
+  const [tempTodo, setTempTodo] = useState<Todo | null>(null);
+  const [updatingIds, setUpdatingIds] = useState<number[]>([]);
+
+  useEffect(() => {
+    getTodos(USER_ID)
+      .then(setTodos)
+      .catch(() => {
+        setErrorMessage(Error.todosLoading);
+
+        return wait(3000).then(() => setErrorMessage(null));
+      });
+  }, []);
+
+  useEffect(() => {
+    wait(3000).then(() => setErrorMessage(Error.none));
+  }, [errorMessage]);
+
+  const filteredTodos = getFilteredTodos(todos, filter);
+
+  const addTodo = useCallback(
+    (newTodo: Todo) => {
+      setTodos(currentTodos => [...currentTodos, newTodo]);
+    },
+    [setTodos],
+  );
+
+  const deleteTodo = useCallback(
+    (todoId: number) => {
+      setTodos(currentTodos =>
+        currentTodos.filter(todo => {
+          return todo.id !== todoId;
+        }),
+      );
+    },
+    [setTodos],
+  );
+
+  const updateTodo = useCallback(
+    ({ title, id, completed }: Omit<Todo, 'userId'>) => {
+      setTodos(currentTodos =>
+        currentTodos.map(currentTodo => {
+          return currentTodo.id !== id
+            ? currentTodo
+            : { ...currentTodo, title, completed };
+        }),
+      );
+    },
+    [setTodos],
+  );
+
+  const handleSetTempTodo = (newTodo: Todo | null) => {
+    setTempTodo(newTodo);
+  };
+
+  const handleSetErrorMessage = (newError: Error) => {
+    setErrorMessage(newError);
+  };
+
+  const handleSetUpdatingIds = (id: number | null) => {
+    if (id) {
+      setUpdatingIds(currentIds => [...currentIds, id]);
+    } else {
+      setUpdatingIds([]);
+    }
+  };
+
+  const filterChange = (newFilter: Status) => {
+    setFilter(newFilter);
+  };
+
+  const todosValue = {
+    todos,
+    filteredTodos,
+    filter,
+    tempTodo,
+    errorMessage,
+    updatingIds,
+    addTodo,
+    deleteTodo,
+    updateTodo,
+    handleSetErrorMessage,
+    handleSetTempTodo,
+    handleSetUpdatingIds,
+    filterChange,
```

Comment:
```suggestion
    filterChange: setFilter,
```
`filterChange` wrapper is redundant

or just
```suggestion
    setFilter,
```